File: control.java
import java.util.Scanner;
import java.util.ArrayList;

public class control {
    Mantenimiento mantenimiento;
    Personal Empleado=new Personal();
    visita visita=new visita(" ",Empleado, 0, 0, 0);
    Scanner leer = new Scanner(System.in);
    Visitantes visitantes = new Visitantes(" ", " ", 0, " ", " ");
    ArrayList<visita> ListaDeVisitas=new ArrayList<>();
    ArrayList<Mantenimiento> listaDeMantenimientos=new ArrayList<>();
    private int buscarEmpleadoPorCURP(String curp) {
        for (int i = 0; i < Empleado.getPersonal().size(); i++) {
            if (curp.equals(Empleado.getPersonal().get(i).getCURP())) {
                return i;
            }
        }
        return -1;
    }
    public void RegistarVisitas() {
        System.out.println("Ingrese la fecha del recorrido con la siguiente estructura: día/mes/año");
        String fechaDeRecorrido = leer.nextLine();

        // Buscar el guía por CURP
        Personal guia = null;
        boolean guiaEncontrado = false;
        while (!guiaEncontrado) {
            System.out.println("Ingrese el CURP del guía que dará el recorrido:");
            String curpGuia = leer.nextLine().toUpperCase();
            int indiceEmpleado = buscarEmpleadoPorCURP(curpGuia);
            if (indiceEmpleado != -1) {
                String rol = Empleado.getPersonal().get(indiceEmpleado).getRol();
                if (rol.equals("GUIA")) {
                    guia = Empleado.getPersonal().get(indiceEmpleado);
                    guiaEncontrado = true;
                } else {
                    System.out.println("Este empleado no puede realizar esa acción.");
                    System.out.println("Si desea cancelar el registro, presione 1; de lo contrario, presione 2:");
                    int opcion = leer.nextInt();
                    if (opcion == 1) {
                        return;
                    }
                }
            } else {
                System.out.println("Empleado no encontrado. Asegúrese de anotar correctamente el CURP.");
                System.out.println("Si desea cancelar el registro, presione 1; de lo contrario, presione 2:");
                int opcion = leer.nextInt();
                if (opcion == 1) {
                    return;
                }
            }
        }
        ArrayList<Visitantes> visitantesEnRecorrido = new ArrayList<>();
        int totalAdultos = 0;
        int totalNiños = 0;
        boolean agregarVisitante = true;
        while (agregarVisitante) {
            System.out.println("Ingrese la CURP del visitante:");
            String curpVisitante = leer.nextLine().toUpperCase();
            int indiceVisitante = visitantes.buscarVisitantePorCURP(curpVisitante);
            if (indiceVisitante != -1) {
                Visitantes visitante = visitantes.getVisitantes().get(indiceVisitante);
                visitante.setNumVisistas(visitante.getNumVisitas() + 1); // Incrementar el número de visitas del visitante
                if (visitante.getNumVisitas() % 5 == 0) { // Verificar si se pueden aplicar descuentos
                    // Reiniciar el contador de visitas
                    visitante.setNumVisistas(0);
                    // Aplicar descuento del 20%
                    System.out.println("¡Felicidades! Se aplicó un descuento del 20% en el boleto para el visitante con CURP: " + visitante.getCURP());
                }
                visitantesEnRecorrido.add(visitante);
                if (visitante.getFechaNacimiento() >= 2012) {
                    totalAdultos++;
                } else {
                    totalNiños++;
                }
            } else {
                System.out.println("Visitante no encontrado.");
            }

            // Preguntar si desea agregar otro visitante
            System.out.println("¿Desea agregar otro visitante? (s/n)");
            String respuesta = leer.nextLine().toLowerCase();
            if (respuesta.equals("n")) {
                agregarVisitante = false;
            }
        }
        double costoTotal = (totalAdultos * 100) + (totalNiños * 50);
        // Aplicar descuento
        int totalVisitasElegiblesDescuento = 0;
        for (Visitantes visitante : visitantesEnRecorrido) {
            if (visitante.getNumVisitas() % 5 == 0) {
                totalVisitasElegiblesDescuento++;
            }
        }

        if (totalVisitasElegiblesDescuento > 0) {
            double descuento = 0.2; // Descuento del 20%
            double montoDescuento = costoTotal * descuento * (totalVisitasElegiblesDescuento / 5); // Calcular monto de descuento
            costoTotal -= montoDescuento; // Aplicar descuento al costo total
            System.out.println("Se aplicó un descuento del 20% al costo total.");
        }
        visita nuevaVisita = new visita(fechaDeRecorrido, guia, costoTotal, totalAdultos, totalNiños);
        nuevaVisita.setListaDeVisitantesEnRecorrido(visitantesEnRecorrido);
        ListaDeVisitas.add(nuevaVisita);

        // Mostrar información de la visita
        System.out.println("Visita registrada exitosamente:");
        System.out.println("Fecha del recorrido: " + nuevaVisita.fechaDeRecorrido);
        System.out.println("Guía a cargo: " + nuevaVisita.guia.getName() + " " + nuevaVisita.guia.getApellido());
        System.out.println("Número total de adultos: " + nuevaVisita.NumDeAdultos);
        System.out.println("Número total de niños: " + nuevaVisita.NumDeNiños);
        System.out.println("Costo total de la visita: $" + nuevaVisita.costoTotal);
    }
    public void mostrarTodasLasVisitas() {
        if (ListaDeVisitas.isEmpty()) {
            System.out.println("No hay visitas registradas.");
        } else {
            System.out.println("Lista de todas las visitas registradas:");
            for (visita v : ListaDeVisitas) {
                System.out.println("Fecha del recorrido: " + v.fechaDeRecorrido);
                System.out.println("Guía a cargo: " + v.guia.getName() + " " + v.guia.getApellido());
                System.out.println("Número total de adultos: " + v.NumDeAdultos);
                System.out.println("Número total de niños: " + v.NumDeNiños);
                System.out.println("Costo total de la visita: $" + v.costoTotal);
                System.out.println("---------------------------------------------");
            }
        }
    }

    public void imprimirInformacionVisita(String fechaDeRegistro) {
        for (visita v : ListaDeVisitas) {
            if (v.fechaDeRecorrido.equals(fechaDeRegistro)) {
                System.out.println("Información de la visita:");
                System.out.println("Fecha del recorrido: " + v.fechaDeRecorrido);
                System.out.println("Guía a cargo: " + v.guia.getName() + " " + v.guia.getApellido());
                System.out.println("Número total de adultos: " + v.NumDeAdultos);
                System.out.println("Número total de niños: " + v.NumDeNiños);
                System.out.println("Costo total de la visita: $" + v.costoTotal);
                System.out.println("Visitantes:");
                for (Visitantes visitante : v.getListaDeVisitantesEnRecorrido()) {
                    System.out.println("Nombre: " + visitante.getNombre());
                    System.out.println("Apellidos: " + visitante.getApellidos());
                    System.out.println("Fecha de nacimiento: " + visitante.getFechaNacimiento());
                    System.out.println("CURP: " + visitante.getCURP());
                    System.out.println("Número de visitas: " + visitante.getNumVisitas());
                    System.out.println("Fecha de registro: " + visitante.getFechaRegistro());
                }
                return;
            }
        }
        System.out.println("No se encontró ninguna visita registrada para la fecha de registro proporcionada.");
    }

    public void registrarMantenimiento() {
        System.out.println("Ingrese el CURP del empleado que realizará el mantenimiento:");
        String curpEmpleado = leer.nextLine().toUpperCase();
        int indiceEmpleado = buscarEmpleadoPorCURP(curpEmpleado);

        if (indiceEmpleado != -1) {
            Personal empleado = Empleado.getPersonal().get(indiceEmpleado);
            if (empleado.getRol().equals("MANTENIMIENTO")) {
                System.out.println("Ingrese el ID del animal:");
                int ID = leer.nextInt();
                leer.nextLine(); // Limpiar el buffer

                System.out.println("Ingrese el proceso realizado:");
                String procesoRealizado = leer.nextLine();

                System.out.println("Ingrese la fecha del proceso (formato: día/mes/año):");
                String fechaProceso = leer.nextLine();

                System.out.println("Ingrese observaciones (opcional):");
                String observaciones = leer.nextLine();

                Mantenimiento mantenimiento = new Mantenimiento(empleado, ID, procesoRealizado, fechaProceso, observaciones);
                listaDeMantenimientos.add(mantenimiento);

                System.out.println("Mantenimiento registrado exitosamente.");
            } else {
                System.out.println("El empleado no tiene permisos para realizar mantenimientos.");
            }
        } else {
            System.out.println("Empleado no encontrado.");
        }
    }


    // Método para imprimir todos los mantenimientos realizados
    public void imprimirMantenimientos() {
        System.out.println("Lista de mantenimientos realizados:");
        for (int i = 0; i < listaDeMantenimientos.size(); i++) {
            System.out.println("Empleado encargado: " + mantenimiento.empleado.getName() + " " + mantenimiento.empleado.getApellido());
            System.out.println("ID del animal: " + mantenimiento.ID);
            System.out.println("Proceso realizado: " + mantenimiento.ProcesoQuSeRealizo);
            System.out.println("Fecha del proceso: " + mantenimiento.fechaDeMantenimento);
            System.out.println("Observaciones: " + (mantenimiento.observaciones.isEmpty() ? "Ninguna" : mantenimiento.observaciones));
            System.out.println("---------------------------------------------");
        }
    }

    // Método para consultar un mantenimiento por fecha
    public void consultarMantenimientoPorFecha(String fecha) {
        boolean encontrado = false;
        for (int i = 0; i < listaDeMantenimientos.size(); i++) {
            if (mantenimiento.fechaDeMantenimento.equals(fecha)) {
                encontrado = true;
                System.out.println("Empleado encargado: " + mantenimiento.empleado.getName() + " " + mantenimiento.empleado.getApellido());
                System.out.println("ID del animal: " + mantenimiento.ID);
                System.out.println("Proceso realizado: " + mantenimiento.ProcesoQuSeRealizo);
                System.out.println("Fecha del proceso: " + mantenimiento.fechaDeMantenimento);
                System.out.println("Observaciones: " + (mantenimiento.observaciones.isEmpty() ? "Ninguna" : mantenimiento.observaciones));
                System.out.println("---------------------------------------------");
            }
        }
        if (!encontrado) {
            System.out.println("No se encontraron mantenimientos realizados en la fecha especificada.");
        }
    }
}

File: control.py
class Control:
    def __init__(self):
        self.mantenimiento = None
        self.empleado = Personal()
        self.visita = Visita("", self.empleado, 0, 0, 0)
        self.leer = Scanner()
        self.visitantes = Visitantes("", "", 0, "", "")
        self.lista_de_visitas = []
        self.lista_de_mantenimientos = []

    def buscar_empleado_por_curp(self, curp):
        for i in range(len(self.empleado.personal)):
            if curp == self.empleado.personal[i].curp:
                return i
        return -1

    def registrar_visitas(self):
        print("Ingrese la fecha del recorrido con la siguiente estructura: día/mes/año")
        fecha_de_recorrido = self.leer.next_line()

        guia = None
        guia_encontrado = False
        while not guia_encontrado:
            print("Ingrese el CURP del guía que dará el recorrido:")
            curp_guia = self.leer.next_line().upper()
            indice_empleado = self.buscar_empleado_por_curp(curp_guia)
            if indice_empleado != -1:
                rol = self.empleado.personal[indice_empleado].rol
                if rol == "GUIA":
                    guia = self.empleado.personal[indice_empleado]
                    guia_encontrado = True
                else:
                    print("Este empleado no puede realizar esa acción.")
                    opcion = self.leer.next_int()
                    if opcion == 1:
                        return
            else:
                print("Empleado no encontrado. Asegúrese de anotar correctamente el CURP.")
                opcion = self.leer.next_int()
                if opcion == 1:
                    return

        visitantes_en_recorrido = []
        total_adultos = 0
        total_niños = 0
        agregar_visitante = True
        while agregar_visitante:
            print("Ingrese la CURP del visitante:")
            curp_visitante = self.leer.next_line().upper()
            indice_visitante = self.visitantes.buscar_visitante_por_curp(curp_visitante)
            if indice_visitante != -1:
                visitante = self.visitantes.visitantes[indice_visitante]
                visitante.num_visitas += 1
                if visitante.num_visitas % 5 == 0:
                    visitante.num_visitas = 0
                    print("¡Felicidades! Se aplicó un descuento del 20% en el boleto para el visitante con CURP:", visitante.curp)
                visitantes_en_recorrido.append(visitante)
                if visitante.fecha_nacimiento >= 2012:
                    total_adultos += 1
                else:
                    total_niños += 1
            else:
                print("Visitante no encontrado.")

            print("¿Desea agregar otro visitante? (s/n)")
            respuesta = self.leer.next_line().lower()
            if respuesta == "n":
                agregar_visitante = False

        costo_total = (total_adultos * 100) + (total_niños * 50)
        total_visitas_elegibles_descuento = sum(1 for visitante in visitantes_en_recorrido if visitante.num_visitas % 5 == 0)

        if total_visitas_elegibles_descuento > 0:
            descuento = 0.2
            monto_descuento = costo_total * descuento * (total_visitas_elegibles_descuento // 5)
            costo_total -= monto_descuento
            print("Se aplicó un descuento del 20% al costo total.")

        nueva_visita = Visita(fecha_de_recorrido, guia, costo_total, total_adultos, total_niños)
        nueva_visita.lista_de_visitantes_en_recorrido = visitantes_en_recorrido
        self.lista_de_visitas.append(nueva_visita)

        print("Visita registrada exitosamente:")
        print("Fecha del recorrido:", nueva_visita.fecha_de_recorrido)
        print("Guía a cargo:", nueva_visita.guia.name, nueva_visita.guia.apellido)
        print("Número total de adultos:", nueva_visita.num_de_adultos)
        print("Número total de niños:", nueva_visita.num_de_niños)
        print("Costo total de la visita: $", nueva_visita.costo_total)

    def mostrar_todas_las_visitas(self):
        if not self.lista_de_visitas:
            print("No hay visitas registradas.")
        else:
            print("Lista de todas las visitas registradas:")
            for v in self.lista_de_visitas:
                print("Fecha del recorrido:", v.fecha_de_recorrido)
                print("Guía a cargo:", v.guia.name, v.guia.apellido)
                print("Número total de adultos:", v.num_de_adultos)
                print("Número total de niños:", v.num_de_niños)
                print("Costo total de la visita: $", v.costo_total)
                print("---------------------------------------------")

    def imprimir_informacion_visita(self, fecha_de_registro):
        for v in self.lista_de_visitas:
            if v.fecha_de_recorrido == fecha_de_registro:
                print("Información de la visita:")
                print("Fecha del recorrido:", v.fecha_de_recorrido)
                print("Guía a cargo:", v.guia.name, v.guia.apellido)
                print("Número total de adultos:", v.num_de_adultos)
                print("Número total de niños:", v.num_de_niños)
                print("Costo total de la visita: $", v.costo_total)
                print("Visitantes:")
                for visitante in v.lista_de_visitantes_en_recorrido:
                    print("Nombre:", visitante.nombre)
                    print("Apellidos:", visitante.apellidos)
                    print("Fecha de nacimiento:", visitante.fecha_nacimiento)
                    print("CURP:", visitante.curp)
                    print("Número de visitas:", visitante.num_visitas)
                    print("Fecha de registro:", visitante.fecha_registro)
                return
        print("No se encontró ninguna visita registrada para la fecha de registro proporcionada.")

    def registrar_mantenimiento(self):
        print("Ingrese el CURP del empleado que realizará el mantenimiento:")
        curp_empleado = self.leer.next_line().upper()
        indice_empleado = self.buscar_empleado_por_curp(curp_empleado)

        if indice_empleado != -1:
            empleado = self.empleado.personal[indice_empleado]
            if empleado.rol == "MANTENIMIENTO":
                print("Ingrese el ID del animal:")
                ID = self.leer.next_int()
                self.leer.next_line()

                print("Ingrese el proceso realizado:")
                proceso_realizado = self.leer.next_line()

                print("Ingrese la fecha del proceso (formato: día/mes/año):")
                fecha_proceso = self.le

class Mantenimiento:
    def __init__(self, empleado, ID, proceso_realizado, proceso_que_se_realizo, fecha_de_mantenimiento):
        self.empleado = empleado
        self.animal = Animales()
        self.ID = ID
        self.observaciones = ""
        self.fecha_de_mantenimiento = fecha_de_mantenimiento
        self.proceso_qu_se_realizo = proceso_que_se_realizo

class Visita:
    def __init__(self, fecha_de_recorrido, guia, costo_total, num_de_adultos, num_de_niños):
        self.lista_de_visitas = []
        self.fecha_de_recorrido = fecha_de_recorrido
        self.guia = guia
        self.visitante = Visitantes("", " ", 0, " ", " ")
        self.costo_total = costo_total
        self.num_de_niños = num_de_niños
        self.num_de_adultos = num_de_adultos
        self.lista_de_visitantes_en_recorrido = []

    def agregar_visitantes_a_recorrido(self, visitante):
        self.lista_de_visitantes_en_recorrido.append(visitante)

    # Getters and setters
    def get_lista_de_visitantes_en_recorrido(self):
        return self.lista_de_visitantes_en_recorrido

    def set_lista_de_visitantes_en_recorrido(self, lista_de_visitantes_en_recorrido):
        self.lista_de_visitantes_en_recorrido = lista_de_visitantes_en_recorrido